### PR TITLE
Zbeekman mac os patch

### PR DESCRIPTION
**Pull request by zbeekman**
I can squash down the commits, etc.

Should help fix issues raised in #4

### CODE REVIEW

**Comment by zbeekman** [comment]
One, moment, this doesn't work...

**Comment by pocc** [comment]
Will merge once it does.

**Comment by zbeekman** [comment]
@pocc: AFAICT, this is now ready to merge. Here are a few notes:

1. Fix for non-robust "$0" directory finding was adapted from both:
   1. https://www.ostricher.com/2014/10/the-right-way-to-get-the-directory-of-a-bash-script/ ("Final attempt: Resolve symlinks in `$BASH_SOURCE`"
   2. http://stackoverflow.com/questions/59895/can-a-bash-script-tell-what-directory-its-stored-in
2. `clang-tidy` on my system was producing a new, unexpected error for `err.cpp`
   1. c055e3ebb8914303caeb924467f0376236e72b7f modifies the test to instruct `clang-tidy` to ignore that particular error, but a more robust solution would be to make the tests less tightly coupled to the output
3. `oclint` on my system has a patch revision (`0.13.1`, NOT `0.13`) which breaks the test. I did not fix this.
3. `oclint` seems to output errors/warnings in a non-deterministic order, and the test breaks when the order changes. I did not fix this either.

Also, on macOS, the homebrew cask for oclint conflicts with GCC (won't install, and causes issues if it is installed and you try to install/link gcc). A fix for this was just merged: https://github.com/Homebrew/homebrew-cask/pull/73649

**Comment by zbeekman** [comment]
@pocc if you want me to drop the test modification (if you have plans for a better fix) or want me to make any changes,  please let me know.

File: hooks/clang-format [hunk 1]
```diff
@@ -3,7 +3,19 @@
 ################################################################################
 CMD="clang-format"
 args=("$@")
-hook_dir="$(dirname "$(readlink -f "$0")")"
+get_script_dir () {
+     SOURCE="${BASH_SOURCE[0]}"
+     # While $SOURCE is a symlink, resolve it
+     while [ -h "$SOURCE" ]; do
+          DIR="$( cd -P "$( dirname "$SOURCE" )" > /dev/null && pwd )"
```
Comment (guihao-liang) [inline hooks/clang-format]: There's no need to keep extra space around parenthesis.

File: hooks/clang-format [hunk 2]
```diff
@@ -3,7 +3,19 @@
 ################################################################################
 CMD="clang-format"
 args=("$@")
-hook_dir="$(dirname "$(readlink -f "$0")")"
+get_script_dir () {
+     SOURCE="${BASH_SOURCE[0]}"
```
Comment (guihao-liang) [inline hooks/clang-format]: `$0` is sufficient.

Comment (guihao-liang) [inline hooks/clang-format]: you may consider making it local, `local source_dir=`

**Comment by guihao-liang** [comment]
Sorry, I didn't mean to be a reviewer.

**Comment by guihao-liang** [comment]
I found this [link](https://stackoverflow.com/questions/5756524/how-to-get-absolute-path-name-of-shell-script-on-macos) is pretty helpful to have the right `readlink` workaround.
 
Besides, that, it's better to use `$OSTYPE` to detect os first, if it's "linux-gnu", use `readlink -f` directly.

**Comment by guihao-liang** [comment]
> eep this function in one place instead of several copies?

nvm, it's a chicken and egg problem.

**Comment by pocc** [comment]
Closing as this no longer necessary. All of the utils functions have been merged into each respective file for this bash version and a python version is now available. @zbeekman thank you for making this PR.

If I'm missing something, please reply here.

**Pull request closed without merging**